### PR TITLE
feat: Add draft for post on SCv2 Hodometer

### PR DESCRIPTION
**Pull request by agrski**
# Why
## Motivation
Hodometer is one of the components in Seldon Core v2.  Any series of blog posts on Core v2 should therefore discuss it: what it is, how it works, why it was designed the way it was and any history or context informing those decisions, and so on.

I've chosen to start by drafting a post on Hodometer because I probably know it and the design decisions the best, and because it's a little independent of most of the rest of the system, thus there shouldn't be too many cross-references to iron out.

## Issues
N/A

# What
## Changes
* Add draft blog post about the SCv2 component Hodometer.

## Misc
In case it's easier to read and review, here is the link to the rendered Markdown: https://github.com/agrski/agrski.github.io/blob/add-draft-for-hodometer-post/docs/_drafts/scv2-hodometer.md

Certain formatting might look odd, like `---`, because that'll be rendered by Jekyll but not Markdown.

### CODE REVIEW

File: docs/_drafts/scv2-hodometer.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+For example, the number of nodes in a cluster has no relation to how many ML deployments or models are on it, or what hardware resources those are consuming.
+It also doesn't provide any information on which version of Core has been installed, or anything else that's application-specific.
+Spartakus does support some custom data in the form of [extensions](https://github.com/kubernetes-retired/spartakus/tree/master/docs#extensions), but this is static information.
+In any case, users may not even want to share all the data on nodes in a cluster that Spartakus collects by default.
+To put it concisely, Spartakus collected lots of data that _wasn't_ of interest but didn't collect lots of data that _was_ of interest!
+
+## Designing Hodometer
+
+Core v2 is a very different kettle of fish from Core v1, and the sorts of metrics we'd want are no exception.
+Core v2 has multiple Kubernetes custom resources --- models, servers, pipelines, and experiments --- compared to just single one in Core v1 --- deployments.
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md]: :thought_balloon: Consider using the exact terms for resources, e.g. `SeldonDeployment`

File: docs/_drafts/scv2-hodometer.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+While Spartakus didn't provide ideal metrics even for Core v1, it did make a number of sensible design decisions to use as a basis for implementing usage metrics in Core v2.
+Specifically, these were:
+* Anonymity via randomised, non-persistent identifiers.
+* Explicitly no collection of PII.
+* Non-integral component.
+* Low periodicity reporting of metrics.
+* Push-based reporting of metrics.
+* Flexibility to configure additional metrics receivers.
+
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md:55]: :thought_balloon: Consider adding quoted text to emphasise specific design considerations and/or their impact

File: docs/_drafts/scv2-hodometer.md
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md:1]: :thought_balloon: Consider adding Mermaid diagrams explaining how various things communicate, e.g. to contrast Spartakus & Hodometer, and to explain Hodometer's internal architecture

File: docs/_drafts/scv2-hodometer.md (hunk 2 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:55]: For context, other blogs and articles use quotes as a form of call-out to emphasise specific sentences from the text

**Review by ukclivecox — COMMENTED** [review]
Seems good but maybe if this was in a blog I assume you would also be addressing a wider audience with no interest in SCV2 so might be "tips and tricks for providing usage monitoring for your open source project" or something?

File: docs/_drafts/scv2-hodometer.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+While Spartakus treated it is an all-or-nothing situation, Core v2 gives the user the option to share only as much information as they're comfortable with.
+Making users select every individual metric would be tedious for them, so Hodometer groups these into levels: cluster-level, resource-level, and feature-level information.
+Each level is a superset of its predecessor.
+Cluster-level metrics are very basic: just the k8s and Core v2 scheduler versions (the scheduler is the heart of the control plane, so this implies the Core v2 version more generally).
+Resource-level metrics count the number of each Core v2-defined _custom resource_, such as the number of models and pipelines.
+It's important to note here that sensitive information like model names is _not_ recorded.
+Feature-level metrics are about how many servers have multi-model serving and over-committing enabled and how much memory servers have available to them.
+All of this can be seen in the `Collect` method [here](https://github.com/SeldonIO/seldon-core/blob/d3502062bbbb18a08032201917ceea07e124be41/hodometer/pkg/hodometer/collect.go#L227).
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md]: :wrench: Should also link to `metrics.go` for the metrics definitions.  The `Collect` method best shows that levels are supersets/subsets of one another.

File: docs/_drafts/scv2-hodometer.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+This led to the creation of the open-source [i-am-spartakus](https://github.com/SeldonIO/i-am-spartakus) repository a few weeks later by Seldon, to retain the project in maintenance mode.
+Plenty of projects are archived, however, and maintaining a fork of something stable isn't much of a burden.
+The larger problem with Spartakus was that the metrics it collected made sense for k8s, but not necessarily for other applications.
+For example, the number of nodes in a cluster has no relation to how many ML deployments or models are on it, or what hardware resources those are consuming.
+It also doesn't provide any information on which version of Core has been installed, or anything else that's application-specific.
+Spartakus does support some custom data in the form of [extensions](https://github.com/kubernetes-retired/spartakus/tree/master/docs#extensions), but this is static information.
+In any case, users may not even want to share all the data on nodes in a cluster that Spartakus collects by default.
+To put it concisely, Spartakus collected lots of data that _wasn't_ of interest but didn't collect lots of data that _was_ of interest!
+
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md:43]: :thought_balloon: Consider adding a system-level diagram for Spartakus to use as a point of comparison for ones about Hodometer later

**Comment by agrski** [comment]
> Seems good but maybe if this was in a blog I assume you would also be addressing a wider audience with no interest in SCV2 so might be "tips and tricks for providing usage monitoring for your open source project" or something?

Interesting suggestion!  The intent for this blog post (I realise I haven't added a PR description yet) is to be about SCv2 specifically, as part of a series of deep-dives into the design and characteristics of its components and the system as a whole.

With that said, it seems like a good suggestion to have a summary at the end about things to consider if designing a similar metrics system for some other project, e.g. push vs. pull data collection, if support for different data stores is important, collection levels, and so on.

It'd also be worth having a bit more of an introduction to those topics towards the start--the good old "tell 'em what you're gonna tell 'em"!

File: docs/_drafts/scv2-hodometer.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+## A blast from the past --- inspiration from Spartakus
+
+Core v2 is a very different kettle of fish from Core v1, and the sorts of metrics we'd want are no exception.
+Core v2 has multiple Kubernetes custom resources --- models, servers, pipelines, and experiments --- compared to just single one in Core v1 --- deployments.
+It furthermore introduces fundamentally new features in the form of multi-model serving (MMS) and over-commit.
+When you consider these things, it should be clear that the metrics for Core v2 will want to be correspondingly more granular and domain-focused.
+
+While Spartakus didn't provide ideal metrics even for Core v1, it did make a number of sensible design decisions to use as a basis for implementing usage metrics in Core v2.
+Specifically, these were:
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md]: :thought_balloon: These might be better handled as level three headings for the paragraphs below.  That would emphasise them better whilst simultaneously splitting the following content up into more visually manageable chunks.

Comment (agrski) [reply docs/_drafts/scv2-hodometer.md]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/ebd6829836084fcbd575c2feb3b50928b18546e8

File: docs/_drafts/scv2-hodometer.md (hunk 1 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/70a3df4d13eb3bae8ee5fe5096002274f15be2ac and https://github.com/agrski/agrski.github.io/pull/40/commits/be55202a0abc2726c49e7abf95e95f599ee1f0dc

**Comment by agrski** [comment]
@ukclivecox Would you mind having a squizz at the diagrams (the PNGs) and confirming those look reasonable?  As a brief explanation, the _container_ diagrams are about deployable units, and the _component_ diagrams are roughly analogous to services/interfaces/modules; they don't reflect how things are wired up and instantiations of those interfaces, as that's what the _code_ level in the C4 model is for.

I need to write it up still, but actually looking into this and creating these diagrams has brought up some interesting similarities and differences between Spartakus and Hodometers that I hadn't noticed before.

File: docs/_drafts/scv2-hodometer.md (hunk 4 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:43]: Diagrams added for both Spartakus & Hodometer at the container and also component levels

File: docs/_drafts/scv2-hodometer.md
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:1]: Diagrams added

File: docs/_drafts/scv2-hodometer.md (hunk 4 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:43]: May still want to add a link to it here, but perhaps it's fine to have that later

File: docs/_drafts/scv2-hodometer.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
+---
+title: "Seldon Core v2 Deep-Dive---Hodometer"
+tags: seldon seldon-core-v2
+---
+
+Hodometer is an optional component of Seldon Core v2 responsible for collecting anonymous usage metrics.
+The name comes from the [ancient device](https://en.wikipedia.org/wiki/Odometer) for measuring distance, familiar in cars as an _odometer_, because it's all about keeping track of your mileage!
+
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md:10]: :wrench: Add section on who might be interested in this article, e.g. those looking to understand Core v2 in greater depth, those looking to implement usage metrics in a responsible manner for their own projects, and those simply looking to understand the design decisions in an (until recently) open-source project.

File: docs/_drafts/scv2-hodometer-hodometer-component.png
Comment (agrski) [inline docs/_drafts/scv2-hodometer-hodometer-component.png:1]: :thought_balloon: Investigate adding copyright notices to the images, as these are liable to being reused by referencing articles and search engines, which may or may not include proper attribution

Comment (agrski) [reply docs/_drafts/scv2-hodometer-hodometer-component.png:1]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/17e59d6d62ad66c4b89d3ecc18179f03f6e0a8fc

For reference, although I'll likely end up adding this into a Makefile at some point for convenience, I used the following bash script to add copyright notices:
```bash
find docs/_drafts/ -name '*.png' | while read -r F; do convert $F -pointsize 12 -fill grey50 label:"Copyright (C) agrski 2024" -gravity center -append $F; done
```

File: docs/_drafts/scv2-hodometer-spartakus-container.mmd [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+  classDef focusSystem      fill:#1168bd,stroke:#0b4884,color:#ffffff
+  classDef supportingSystem fill:#666,stroke:#0b4884,color:#ffffff
+  classDef consumingSystem  fill:#08427b,stroke:#052e56,color:#ffffff
+
+  class S focusSystem
+  class C consumingSystem
+  class K supportingSystem
+
+  style k8s-cluster fill:#eee,stroke:#ccc,stroke-width:2px,stroke-dasharray: 5 5
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer-spartakus-container.mmd]: :question: Can subgraph styles be turned into a `classDef`?  If so, that would slightly simplify the definitions by providing better grouping and supporting multiple subgraphs (if required in the future) without duplication

Comment (agrski) [reply docs/_drafts/scv2-hodometer-spartakus-container.mmd]: The answer is a resounding (and fortunately unsurprising) yes.  Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/586cee78924a3b247ee4fe4734c9a81882f4849d

File: docs/_drafts/scv2-hodometer.md (hunk 3 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/7c515e074dd451549955c4ca18232b5c118ec76f

File: docs/_drafts/scv2-hodometer.md (hunk 2 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:55]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/be4b17b10133bf35dac2b82ea592aaa83176e137

File: docs/_drafts/scv2-hodometer.md (hunk 6 above)
Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:10]: @ukclivecox Once I've resolved this comment, that should also address your own previous comment.  Aside from that, I think this nearing completion now as a draft.  (A draft in the Jekyll sense, i.e. that it isn't "published" on the blog, not that it isn't in a good state.)

Comment (agrski) [reply docs/_drafts/scv2-hodometer.md:10]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/1ece90a8c3c9ec37389da49be94c16d5d946adc0

**Comment by agrski** [comment]
One final point to consider is ensuring consistency around formatting, i.e. the semantics for preformatted/monospace text, italics, and bold-face:
* Monospace should be for elements taken directly from code, e.g. file, struct, and method names.
* Bold should be used to highlight key terms that are not part of a code-base, e.g. the names of C4 diagram levels.
* Italics should be used for general emphasis in all other contexts, e.g. contrasting trade-offs between systems.

---

Edit: fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/b5587744a560143310c1ba05cf81a0c761d9f15c, https://github.com/agrski/agrski.github.io/pull/40/commits/16ea0e98dcda0ae6693bfc44df6e7e90bafea6d0, and https://github.com/agrski/agrski.github.io/pull/40/commits/4ba5c118922bf76fd3d0ae4fcb732405ff8afd59.

File: docs/_drafts/scv2-hodometer-hodometer-container.mmd [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  end
+  H ----> |"pushes metrics to\n[HTTP]"| R1
+  H -...-> |"pushes metrics to\n[HTTP]"| R3
+
+  classDef focusSystem      fill:#1168bd,stroke:#0b4884,color:#ffffff
+  classDef supportingSystem fill:#666,stroke:#0b4884,color:#ffffff
+  classDef consumingSystem  fill:#08427b,stroke:#052e56,color:#ffffff
+  classDef systemBoundary   fill:#eee,stroke:#ccc,stroke-width:2px,stroke-dasharray: 5 5
+
+  class H,S,K focusSystem
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer-hodometer-container.mmd]: :wrench: K and S are supporting systems, not focus systems

Comment (agrski) [reply docs/_drafts/scv2-hodometer-hodometer-container.mmd]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/86476803ccb4c3acbc9cd65ff33e166b79e1007d

File: docs/_drafts/scv2-hodometer.md [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+It has three main business logic structs: a punctuator, a collector, and a publisher.
+
+![Hodometer component architecture](./scv2-hodometer-hodometer-component.png)
+
+The `Punctuator` struct is responsible for running code on a periodic basis, much like the idea of a punctuator in Kafka Streams, as it happens.
+Unlike the `volunteer` in Spartakus, the `Punctuator` is _not_ responsible for defining what to run.
+Instead, in Hodometer this is provided by the wiring logic in the `main` function, which can be found [here](https://github.com/SeldonIO/seldon-core/blob/d3502062bbbb18a08032201917ceea07e124be41/hodometer/cmd/hodometer/main.go#L70):
+
+```go
+	punctuator.Run(
```
Comment (agrski) [inline docs/_drafts/scv2-hodometer.md]: :wrench: Although correctly indented for where it was copied from the code, the indentation in these code blocks isn't important for readers of the article.  In fact, given the rendering of tabs can be significant (8 spaces in GitHub), it probably hurts legibility to leave the tabs in place.

As such, the tabs in code blocks should be removed.

Comment (agrski) [reply docs/_drafts/scv2-hodometer.md]: Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/748e359c0099285a39080342694949c6478a6da0

**Pull request merged**